### PR TITLE
Send whole transaction pool upon subscription to gossip

### PR DESCRIPTION
**Pull request by AurelienFT**
## Linked Issues/PRs
Part of https://github.com/FuelLabs/fuel-core/issues/1049

## Description
This PR introduce a new workflow when we successfully subscribed to the TX topic in gossip we try ask for all the tx_ids in the pool's peer and then ask the full transactions that we don't have.
This PR also build the `SharedState` of `P2PService` independently to give more flexibility to the initialisation and avoid cyclic dependencies.

### Detailed workflow : 

- When a peer has subscribed to our topic send inform the TxPool
- The TxPool asks the tx ids of the peer
- The TxPool filter the one it already knows and ask for the full transactions of the others (a multi message send should be done if the number of txs asked is above the limit (currently 10k chosen randomly)).
- We verify and add them to our txpool

### Questions for review

- What's happens if the peer doesn't answer to a request that we await ?
- I placed Option in the request/response because there was in the other messages but why it's needed ?

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests
- [x] [The specification](https://github.com/FuelLabs/fuel-specs/) matches the implemented behavior (link update PR if changes are needed)

### Before requesting review
- [x] I have reviewed the code myself
- [x] I have created follow-up issues caused by this PR and linked them here

### CODE REVIEW

**Comment by AurelienFT** [comment]
@rymnc The PR is still in draft because I still need to make tests but you can start take a look the logic shouldn't change much.

**Comment by rymnc** [comment]
> @rymnc The PR is still in draft because I still need to make tests but you can start take a look the logic shouldn't change much.
> 
> Thankssss <3

will look at it deeply tomorrow, hope that works!

**Comment by rymnc** [comment]
looking good, however, I don't think we can call this "gossip" since its a direct p2p request/response protocol given that we make the request to the peer directly (`peer_id`) as part of the request.

> What's happens if the peer doesn't answer to a request that we await ?

we should have a timeout for waiting for the response, and then try another peer (cc: @xgreenx for confirmation)

> I placed Option in the request/response because there was in the other messages but why it's needed ?

for the other data we're sending over the req/res protocol (`sealed_block_headers` and `transactions`), there is a chance that the range provided is greater than the block height, and to handle that we send `None` for all blocks we don't have data for. We don't need Option usage for transaction pool transactions :)

File: crates/services/p2p/src/service.rs
Comment (rymnc) [inline crates/services/p2p/src/service.rs:480]: I think we should prefix the `RequestMessage`'s with `TxPool` to disambiguate between database transactions and txpool transactions :)

**Comment by AurelienFT** [comment]
@rymnc Thanks for your overview and your comments 

About : 

> I don't think we can call this "gossip" since its a direct p2p request/response protocol

I don't think we ever call the request/response gossip, it's just that it's been trigger on the tx gossip topic being subscribed.

>  We don't need Option usage for transaction pool transactions :)

Will remove that

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs]: Done

File: crates/fuel-core/src/service/adapters/txpool.rs
Comment (rymnc) [inline crates/fuel-core/src/service/adapters/txpool.rs:125]: maybe you want to import TxId at the top given you're using the fully qualified type in two places

File: crates/fuel-core/src/service/adapters/txpool.rs
Comment (rymnc) [inline crates/fuel-core/src/service/adapters/txpool.rs:106]: can this be `fuel_core_services::stream::pending()` ?

File: crates/fuel-core/src/service/adapters/txpool.rs
Comment (rymnc) [inline crates/fuel-core/src/service/adapters/txpool.rs:193]: if we wrapped the vec in an option, we could just do `Ok(None)`. perhaps this is better for performance, but this will never be called if p2p isn't enabled, so it doesn't matter i guess

File: crates/services/p2p/src/p2p_service.rs
Comment (rymnc) [inline crates/services/p2p/src/p2p_service.rs:1630]: imo, this is a lot of nesting within the same test, can we abstract some commonly used funcionality out?

File: crates/services/p2p/src/service.rs
Comment (rymnc) [inline crates/services/p2p/src/service.rs:1063]: can we move the magic numbers out of this function and define it as -
a. a global, or
b. include it in the arguments to the function?

**Review by rymnc — COMMENTED** [review]
In general LGTM, added some comments/questions :)

File: crates/fuel-core/src/service/adapters/txpool.rs
Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/txpool.rs:193]: This doesn't change much as it's never called and will bring a lot of overhead everywhere in the code to manage the option

Comment (rymnc) [reply crates/fuel-core/src/service/adapters/txpool.rs:193]: agreed!

File: crates/fuel-core/src/service/adapters/txpool.rs
Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/txpool.rs]: Done :)

File: crates/fuel-core/src/service/adapters/txpool.rs
Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/txpool.rs]: Done 👍

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs]: The code was using magic number, idk why it's not in the config so I added a global.

File: crates/fuel-core/src/service/adapters/p2p.rs
Comment (rymnc) [inline crates/fuel-core/src/service/adapters/p2p.rs:59]: nit: can we have a top level import for TxId please?
```suggestion
    fn get_all_tx_ids(&self) -> Vec<TxId> {
```

File: crates/fuel-core/src/service/adapters/p2p.rs
Comment (rymnc) [inline crates/fuel-core/src/service/adapters/p2p.rs:65]:
```suggestion
        tx_ids: Vec<TxId>,
```

File: crates/fuel-core/src/service/adapters/p2p.rs
Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/p2p.rs]: Done

File: crates/fuel-core/src/service/adapters/p2p.rs
Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/p2p.rs]: Done

**Review by rymnc — APPROVED** [review]

File: crates/services/txpool/src/ports.rs
Comment (Voxelot) [inline crates/services/txpool/src/ports.rs:49]: should we use a PeerId newtype for Vec<u8>?

Comment (AurelienFT) [reply crates/services/txpool/src/ports.rs:49]: I took example from :
https://github.com/FuelLabs/fuel-core/blob/95b373e80094b09ab5dc6c1e351deccf68ff7dd9/crates/services/p2p/src/service.rs#L842

I was thinking that this was the way you used `PeerId` outside of the P2P module. But I just saw that the PeerId type is on the type shared crate I will change this.

File: crates/fuel-core/src/service/adapters/txpool.rs
Comment (Dentosal) [inline crates/fuel-core/src/service/adapters/txpool.rs:129]: Why we just log the error but do not propagate it upwards?

Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/txpool.rs:129]: I don't want the `txpool` service to stop because of the error if it's just that we can't communicate with this peer anymore. Maybe this function shouldn't return a result at all.

Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/txpool.rs]: Done

File: crates/services/txpool/src/ports.rs
Comment (AurelienFT) [reply crates/services/txpool/src/ports.rs]: Done

File: tests/tests/tx_gossip.rs
Comment (Dentosal) [inline tests/tests/tx_gossip.rs:202]: Looks like a [time dependency port](github.com/FuelLabs/fuel-core/pull/2145) would be a good thing to add here as well. As a follow-up, not in this PR.

File: crates/fuel-core/src/service/adapters/txpool.rs
Comment (rymnc) [reply crates/fuel-core/src/service/adapters/txpool.rs]: well, this function should return an error if sending the request to the peer fails, so I don't think we should remove the result from this.

Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/txpool.rs]: Done

**Review by rymnc — APPROVED** [review]

File: CHANGELOG.md
Comment (rymnc) [inline CHANGELOG.md:10]: perhaps this should go after and not before the other entries :)

Comment (AurelienFT) [reply CHANGELOG.md:10]: Let's see when we will merge it it will propbably change again

File: crates/services/p2p/src/service.rs
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:389]:
```suggestion
    tx_pool: T,
    max_headers_per_request: usize,
```

Why do we need dead code here?

File: crates/services/p2p/src/service.rs
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:405]:
```suggestion
impl<V, T> UninitializedTask<V, SharedState, T> {
```

File: crates/services/p2p/src/service.rs
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:429]:
```suggestion
impl<P: TaskP2PService, V, B: Broadcast, T> Task<P, V, B, T> {
```

File: crates/services/p2p/src/service.rs
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:854]: This code currently works for all possible subscribers. I think we just want to know about subscribers for transactions. I think we either check the tag of subscription, or pass the tag along with peer id.

File: crates/services/p2p/src/service.rs
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:984]:
```suggestion
        debug_assert_eq!(
            peer_id.as_ref(),
            response_from_peer.to_bytes(),
            "Bug: response from non-requested peer"
        );
```

File: crates/services/p2p/src/service.rs
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:1005]:
```suggestion
        debug_assert_eq!(
```

File: crates/services/txpool/src/service.rs
Comment (xgreenx) [inline crates/services/txpool/src/service.rs:277]: I think we need to do this change in this PR since this branch of Tokyo select can be very heavy and long. I think we need spawn a task for it.

We can create queue of task by using rayon for example. The same like we did with `crates/services/p2p/src/heavy_task_processor.rs` just to process this use case.

If you want, it can be another PR on top of this one, we will review and merge into this branch

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs]: Done

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs]: Done

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs]: Done

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs]: Outdated spec. Removed.

File: CHANGELOG.md
Comment (AurelienFT) [reply CHANGELOG.md]: Done

File: crates/fuel-core/src/service/adapters/p2p.rs
Comment (xgreenx) [reply crates/fuel-core/src/service/adapters/p2p.rs:68]: I see. Yeah, we can solve it by using `Arc<Transaction>` on ports and adapter levels. Instead of using `Arc<PoolTransaction>` in the TxPool, we could use `struct PoolTransaction(Arc<Transaction>)`. Then, in this case, we could easily get `Arc<Transaction>`.

It is okay to do this change in a separate PR. It would be nice to make this change before merging this PR, if possible. But it is up to you=)

Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/p2p.rs:68]: I can change this in the transaction pool v2. Can we wait tx pool v2 before merging this ?

File: crates/services/txpool/src/service.rs
Comment (AurelienFT) [reply crates/services/txpool/src/service.rs:277]: Why do we we want to use a rayon dedicated thread pool instead of spawning a call tokio task ?

Comment (xgreenx) [reply crates/services/txpool/src/service.rs:277]: We can spawn tokyo task as well, we need to limit the number of active tasks to be one. Either we can write own queue of tasks and manage them, or reuse `rayon::spawn_fifo`.

File: crates/fuel-core/src/service/adapters/p2p.rs
Comment (xgreenx) [reply crates/fuel-core/src/service/adapters/p2p.rs:68]: I think it is better to do in a separate small PR. This branch and TxPoolV2 could use this branch as a base for this change

File: crates/services/txpool/src/service.rs
Comment (AurelienFT) [reply crates/services/txpool/src/service.rs]: Implemented in : https://github.com/FuelLabs/fuel-core/pull/2131/commits/268d794bb0a70f0ccde2e0ad807f5a997c867538

File: crates/fuel-core/src/service/adapters/p2p.rs
Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/p2p.rs]: After discussion I decided to go for a special type in P2P that allow to serialize a pool transaction or simple transaction as transaction and deserialize only in transactiton

Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/p2p.rs]: Here is a diff showing the changes that implement this : https://github.com/FuelLabs/fuel-core/compare/268d794...5e16d87

File: crates/services/txpool/src/service.rs
Comment (xgreenx) [inline crates/services/txpool/src/service.rs:320]: I think you can just call `SharedState::insert`

File: crates/services/p2p/src/service.rs
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:1024]: It would be nice to separate iterator part with the `Ok` part.

Also, it would be better to avoid panic, even if it is not possible in the current code. Or implement `From<NetworkableTransactionPool>` to `Transaction`. Another question, is that possible to have one type for serialization and another type for deseriliazation? Maybe we can avoid this check at all

File: crates/services/p2p/src/service.rs
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:1007]: It would be nice to use `try_send`

File: crates/services/p2p/src/service.rs
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:977]: It would be nice to use `try_send`

File: crates/services/p2p/src/service.rs
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:628]: Potentially TxPool can be very busy with current work and answering can take a lot of time. It is better to use `handle_request` as well

**Review by xgreenx — COMMENTED** [review]
I have small pack of comments, but the change is almost done!=)

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs]: Done

File: crates/fuel-core/src/service/adapters/p2p.rs
Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/p2p.rs:70]: I just want to re-use an existitng method and not make another one that do the same but `.map()` inside the txpool method

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs:628]: I don't really know what you mean by your first sentence. Also in your second setence you want me to use the thread pool to make the ask to the TxPool ? (I can't directly reuse `handle_request` as it's specifically done for range asks (I can rework the function)

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs]: Done.

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs]: Done.

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs]: I have changed to propagate the result.

I didn't find any way to serialize in a type and deserialize to another with the libp2p enum and serde that's why I ended up with this enum.

File: crates/services/txpool/src/service.rs
Comment (AurelienFT) [reply crates/services/txpool/src/service.rs]: Yeah good catch I was duplicating the verif. All the boilerplate and generics makes it very hard to read.

File: crates/services/txpool/src/service.rs
Comment (AurelienFT) [reply crates/services/txpool/src/service.rs]: Done.

File: crates/services/txpool/src/service.rs
Comment (AurelienFT) [reply crates/services/txpool/src/service.rs]: Done

File: crates/services/p2p/src/config.rs
Comment (AurelienFT) [reply crates/services/p2p/src/config.rs]: Done.

File: crates/fuel-core/src/service/adapters/txpool.rs
Comment (AurelienFT) [reply crates/fuel-core/src/service/adapters/txpool.rs]: Done.

File: bin/fuel-core/src/cli/run/p2p.rs
Comment (AurelienFT) [reply bin/fuel-core/src/cli/run/p2p.rs]: Changed. We need to becareful that this value is under the maximum size of the pool as right now there is no "send in parts" protocol.

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs]: Done in https://github.com/FuelLabs/fuel-core/pull/2131

File: crates/services/p2p/src/service.rs
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:1070]: Maybe we need to not send more than `self.max_txs_per_request` as well=)

**Review by xgreenx — APPROVED** [review]

File: crates/services/p2p/src/ports.rs
Comment (AurelienFT) [reply crates/services/p2p/src/ports.rs]: Done.

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs:1081]: Done

Comment (xgreenx) [reply crates/services/p2p/src/service.rs:1081]: I meant it would be nice from TxPool not to request more than our own config allows=)

**Review by xgreenx — DISMISSED** [review]

File: crates/services/p2p/src/service.rs
Comment (AurelienFT) [reply crates/services/p2p/src/service.rs:1082]: Oopsie, Changed.

**Review by xgreenx — APPROVED** [review]

**Review by rymnc — APPROVED** [review]

**Comment by Voxelot** [comment]
Do we reconcile the txs before or after subscribing? If we start processing gossipped txs while we are still syncing the txpool, we may end up dropping them if they depend on earlier txs that havent been synced yet.

**Comment by AurelienFT** [comment]
> Do we reconcile the txs before or after subscribing? If we start processing gossipped txs while we are still syncing the txpool, we may end up dropping them if they depend on earlier txs that havent been synced yet.

The trigger to make this reconciliation is the subscription so yes it happens just after suscribing. You are right it could lead to invalid gossiped transactions from the point of view of the receiver.
Do you have an idea of when we should make the trigger ?
Because if we activate the gossip after the reconciliation there the possibility that we have miss a child during a period where the gossip wasn't activate too.

Possible solution that I see is, given the tx pool v2 is that all gossiped transactions are queued until we finished at least one peer reconcialition and then we start to process them.
It's pretty easy to do given the fact that in TxPool v2 the verif and insert is done in parrallel manged by a threadpool and a queue

**Pull request merged**
